<h1 style="text-align:center; color:#08519c; margin-bottom:30px;">
  <b>Capstone Project</b>
</h1>

<h2 style="text-align:center; color:#023047; margin-top:16px; margin-bottom:20px;">
  <b>"From Data to Dollars: Understanding Salary Determinants"</b>
</h2>

<hr style="border:2px solid #08519c; margin-top:10px; margin-bottom:20px;">



<h3 style="color:#08519c; margin-bottom:0;">🎯 Project Objective</h3>


This project explores how factors such as <span style="color:#08519c"><b>Job Title</b></span>, 
<span style="color:#08519c"><b>Years of Experience</b></span>, 
<span style="color:#08519c"><b>Education Level</b></span>, 
<span style="color:#08519c"><b>Skills</b></span>, 
<span style="color:#08519c"><b>Industry</b></span>, 
<span style="color:#08519c"><b>Company Size</b></span>, 
<span style="color:#08519c"><b>Certifications</b></span>, and 
<span style="color:#08519c"><b>Remote Work</b></span> influence salary outcomes.

<hr style="border:2px solid #08519c; margin-top:4px; margin-bottom:12px;">

### <span style="color:#08519c">📑 Table of Contents</span>  
1️⃣ Load the dataset  

2️⃣ Initial overview (structure)  
    Summary of Initial Overview (Pre‑Cleaning)  

3️⃣ Data Pre‑processing (missing values, encoding, features)  
    Summary of Data Pre‑processing  

4️⃣ Exploratory analysis (patterns, correlations, trends)  
    Final EDA Summary  

5️⃣ Visualizations (clear, recruiter‑ready charts)  

6️⃣ Insights & conclusions

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

<h2 style="color:#08519c; border-bottom:3px solid #08519c; padding-bottom:4px; margin-top:15px;">
  1. Dataset Loading
</h2>

In [3]:
data= pd.read_csv("Job Salary Prediction.csv")

<h2 style="color:#08519c; border-bottom:3px solid #08519c; padding-bottom:4px; margin-top:15px;">
  2. Initial Overview
</h2>

#### <span style="color:#08519c">2.1 Number of Rows and Columns</span>

In [4]:
rows, columns= data.shape
print(f"The dataset contains {rows} rows and {columns} columns.\n")

The dataset contains 250194 rows and 10 columns.



#### <span style="color:#08519c">2.2 Column names</span>

In [5]:
print(data.columns)

Index(['job_title', 'experience_years', 'education_level', 'skills_count',
       'industry', 'company_size', 'location', 'remote_work', 'certifications',
       'salary'],
      dtype='object')


#### <span style="color:#08519c">2.3 Data types of each column</span>

In [7]:
print("\nData types:")
print (data.dtypes)


Data types:
job_title            object
experience_years    float64
education_level      object
skills_count        float64
industry             object
company_size         object
location             object
remote_work          object
certifications      float64
salary                int64
dtype: object


#### <span style="color:#08519c">2.4 Dataset Structure and Basic Checks</span>

In [9]:
data.head()

,job_title,experience_years,education_level,skills_count,industry,company_size,location,remote_work,certifications,salary
0,AI Engineer,10.0,Bachelor,NaN,Healthcare,Medium,India,Hybrid,2.0,109413
1,Data Analyst,5.0,Bachelor,17.0,Telecom,Small,Australia,No,0.0,93764
2,Frontend Developer,18.0,PhD,4.0,Media,NaN,Singapore,No,1.0,148123
3,Business Analyst,19.0,PhD,13.0,Retail,Medium,Canada,Yes,0.0,189123
4,Product Manager,15.0,Bachelor,7.0,Manufacturing,Large,Sweden,Yes,0.0,165069


In [10]:
data.tail()

,job_title,experience_years,education_level,skills_count,industry,company_size,location,remote_work,certifications,salary
250189,Business Analyst,17.0,Master,2.0,Retail,Medium,Netherlands,No,3.0,139419
250190,AI Engineer,16.0,Master,14.0,Healthcare,Startup,UK,Hybrid,4.0,194819
250191,Software Engineer,16.0,PhD,18.0,Education,Small,Australia,Hybrid,3.0,156195
250192,Business Analyst,18.0,Bachelor,5.0,Technology,Small,Sweden,Yes,1.0,116535
250193,Data Scientist,13.0,Diploma,17.0,Telecom,Small,UK,Hybrid,0.0,151402


In [11]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250194 entries, 0 to 250193
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   job_title         250194 non-null  object 
 1   experience_years  250192 non-null  float64
 2   education_level   250175 non-null  object 
 3   skills_count      250193 non-null  float64
 4   industry          250187 non-null  object 
 5   company_size      250175 non-null  object 
 6   location          250181 non-null  object 
 7   remote_work       250191 non-null  object 
 8   certifications    250168 non-null  float64
 9   salary            250194 non-null  int64  
dtypes: float64(3), int64(1), object(6)
memory usage: 19.1+ MB
None


In [12]:
print("\nSummery statistics:")
data.describe()


Summery statistics:


,experience_years,skills_count,certifications,salary
count,250192.000000,250193.000000,250168.000000,250194.000000
mean,10.005092,9.997742,2.491885,145716.230825
std,6.060777,5.479552,1.706366,37407.830337
min,0.000000,1.000000,0.000000,31867.000000
25%,5.000000,5.000000,1.000000,119356.000000
50%,10.000000,10.000000,2.000000,143452.500000
75%,15.000000,15.000000,4.000000,169491.750000
max,20.000000,19.000000,5.000000,333046.000000


<h3 style="color:#08519c; margin-bottom:0;">🧾 Summary of Initial Overview (Pre‑Cleaning)</h3>
<hr style="border:2px solid #08519c; margin-top:4px; margin-bottom:12px;">

- **Dataset size:** 250,194 rows × 10 columns  
- **Data types:** 6 categorical, 4 numerical  
- **Missing values:** Minor gaps in predictors, salary complete  
- **Key findings:**  
  - Experience averages ~10 years (0–20)  
  - Skills count averages ~10 (1–19)  
  - Certifications mostly 2–3 (0–5)  
  - Salary mean ~145K, ranging from ~31K to ~333K  
- **Overall:** Large, well‑structured dataset with slight missing data and wide salary variation

<h2 style="color:#08519c; border-bottom:3px solid #08519c; padding-bottom:4px; margin-top:15px;">
  3. Data Pre-processing
</h2>

#### <span style="color:#08519c">3.1  Handling missing values</span>

##### <font color='#0891b2'>Checking Missing Values:

In [13]:
print(data.isnull().sum())

job_title            0
experience_years     2
education_level     19
skills_count         1
industry             7
company_size        19
location            13
remote_work          3
certifications      26
salary               0
dtype: int64


##### <font color='#0891b2'>For Numerical columns Handling missing values using Median: 

In [14]:
data['experience_years'] = data['experience_years'].fillna(data['experience_years'].median())

data['skills_count']     = data['skills_count'].fillna(data['skills_count'].median())

data['certifications']   = data['certifications'].fillna(data['certifications'].median())

##### <font color='#0891b2'>For Categorical columns Handling missing values using Mode:

In [15]:
data['education_level']  = data['education_level'].fillna(data['education_level'].mode()[0])

data['industry']         = data['industry'].fillna(data['industry'].mode()[0])

data['company_size']     = data['company_size'].fillna(data['company_size'].mode()[0])

data['location']         = data['location'].fillna(data['location'].mode()[0])

data['remote_work']      = data['remote_work'].fillna(data['remote_work'].mode()[0])

##### <font color='#0891b2'>Checking missing values are resolved:

In [16]:
print("Missing values after handling:\n", data.isnull().sum())

Missing values after handling:
 job_title           0
experience_years    0
education_level     0
skills_count        0
industry            0
company_size        0
location            0
remote_work         0
certifications      0
salary              0
dtype: int64


#### <span style="color:#08519c">3.2  Removing Duplicates</span>

##### <font color='#0891b2'>checking for duplicate rows:

In [18]:
duplicates_count= data.duplicated().sum()
print("Number of duplicate rows:",duplicates_count)

Number of duplicate rows: 194


##### <font color='#0891b2'>Removing duplicate rows:

In [19]:
data=data.drop_duplicates()

##### <font color='#0891b2'>Checking duplicate values resolved:

In [21]:
duplicates_count= data.duplicated().sum()
print("Number of duplicate rows:",duplicates_count)

Number of duplicate rows: 0


#### <span style="color:#08519c">3.3  Correcting data types</span>

##### <font color='#0891b2'>Converting numerical columns to integers:

In [23]:
data['experience_years'] = data['experience_years'].astype('int64')
data['skills_count']     = data['skills_count'].astype('int64')
data['certifications']   = data['certifications'].astype('int64')

##### <font color='#0891b2'>Converting categorical columns to 'category':

In [24]:
categorical_cols = ['job_title', 'education_level', 'industry',
                    'company_size', 'location', 'remote_work']

for col in categorical_cols:
    data[col] = data[col].astype('category')

##### <font color='#0891b2'>Verifying changes:

In [25]:
print(data.dtypes)

job_title           category
experience_years       int64
education_level     category
skills_count           int64
industry            category
company_size        category
location            category
remote_work         category
certifications         int64
salary                 int64
dtype: object


#### <span style="color:#08519c">3.4  Creating derived columns</span>

##### <font color='#0891b2'>Experience group

In [26]:
data['experience_group'] = pd.cut(
    data['experience_years'],
    bins=[0,5,10,15,20],
    labels=['Junior','Mid','Senior','Expert'])

##### <font color='#0891b2'>Skill level

In [27]:
data['skill_level'] = pd.cut(
    data['skills_count'],
    bins=[0,5,10,15,20],
    labels=['Low','Medium','High','Very High'])

##### <font color='#0891b2'>Certification Status 

In [30]:
data['has_certification'] = (data['certifications'] > 0).astype(int)

##### <font color='#0891b2'>Work Mode

In [28]:
data['work_mode'] = data['remote_work'].map({
    'Yes': 'Remote',
    'No': 'On-Site',
    'Hybrid': 'Hybrid'
})

##### <font color='#0891b2'>Salary band

In [29]:
data['salary_band'] = pd.cut(
    data['salary'],
    bins=[0,100000,150000,200000,350000],
    labels=['Low','Mid','High','Very High'])

##### <font color='#0891b2'>Verifying New columns added:

In [31]:
print(data.columns)

Index(['job_title', 'experience_years', 'education_level', 'skills_count',
       'industry', 'company_size', 'location', 'remote_work', 'certifications',
       'salary', 'experience_group', 'skill_level', 'work_mode', 'salary_band',
       'has_certification'],
      dtype='object')


#### <span style="color:#08519c">3.5  Filtering and aggregating data</span>

##### <font color='#0891b2'>Aggregating: count of employees by job title

In [32]:
job_counts = data['job_title'].value_counts()

print(job_counts)

job_title
Backend Developer            21125
Cybersecurity Analyst        20959
Product Manager              20950
AI Engineer                  20945
Data Scientist               20890
DevOps Engineer              20889
Software Engineer            20876
Data Analyst                 20722
Cloud Engineer               20686
Machine Learning Engineer    20677
Business Analyst             20648
Frontend Developer           20633
Name: count, dtype: int64


##### <font color='#0891b2'>Average salary by education level

In [33]:
edu_salary = data.groupby('education_level', observed=False)['salary'].mean().round(2)
print(edu_salary)

education_level
Bachelor       142410.66
Diploma        137160.31
High School    131715.56
Master         153299.45
PhD            163975.93
Name: salary, dtype: float64


In [34]:
data_cleaned= data.copy()

### <font color='#08519c'> Final Data Preview </font>


After completing all preprocessing steps (cleaning, type corrections, derived columns, filtering, aggregation), here’s the updated dataset:

In [35]:
data_cleaned.head()

,job_title,experience_years,education_level,skills_count,industry,company_size,location,remote_work,certifications,salary,experience_group,skill_level,work_mode,salary_band,has_certification
0,AI Engineer,10,Bachelor,10,Healthcare,Medium,India,Hybrid,2,109413,Mid,Medium,Hybrid,Mid,1
1,Data Analyst,5,Bachelor,17,Telecom,Small,Australia,No,0,93764,Junior,Very High,On-Site,Low,0
2,Frontend Developer,18,PhD,4,Media,Small,Singapore,No,1,148123,Expert,Low,On-Site,Mid,1
3,Business Analyst,19,PhD,13,Retail,Medium,Canada,Yes,0,189123,Expert,High,Remote,High,0
4,Product Manager,15,Bachelor,7,Manufacturing,Large,Sweden,Yes,0,165069,Senior,Medium,Remote,High,0


In [36]:
rows, columns= data_cleaned.shape
print(f"The dataset contains {rows} rows and {columns} columns.\n")

The dataset contains 250000 rows and 15 columns.



In [37]:
# Save the cleaned dataset as a CSV
data_cleaned.to_csv("data_cleaned.csv", index=False)

<h3 style="color:#08519c; margin-bottom:0;">🧾 Summary of Data Pre‑processing</h3>
<hr style="border:2px solid #08519c; margin-top:4px; margin-bottom:12px;">

- **Cleaning:** Filled missing values with median; removed duplicates  
- **New Features:** Salary range category, experience bracket  
- **Filtering & Aggregation:** Employee counts by job title; average salary by education  
- **Current Dataset:** 250,000 rows × 16 columns  